### PR TITLE
Community mode

### PR DESCRIPTION
**Pull request by fuentesloic**

### CODE REVIEW

**Comment by fuentesloic** [comment]
@vinyll , @newick could you check all my stories and tell me if I missed something ?

@vinyll I write some python, yes I know that you are thinking so stop ^^, could you check it ...
I still have the Circle CI test wrong because I merged here `privacy location names (#174)` wich have a probleme I think. I don't know how handle it.

1. A Neomad could know if he is in a private mode (opposite of community mode) or not directly on `/profile`

capture d ecran 2017-09-20 a 17 32 30
--------------
capture d ecran 2017-09-20 a 17 32 00



2. A Neomad could enable his profile in `/privacy` from private to community mode with informations wich will affect him privacy/data

capture d ecran 2017-09-20 a 17 01 06

3. A Neomad could disable his profile in `/privacy` from community to private mode with informations wich will affect him privacy/data

capture d ecran 2017-09-20 a 17 01 14

4. A visitor can NOT see datas about a neomad  even if the visitor had the good path to find him
*upgrade* --- need to use the username to define the neomad (and it's in the url so no problems for me to share only this data.

capture d ecran 2017-09-20 a 17 05 01

5. A Neomad can still use all the pages even I choose to not share with the community mode

6. A Neomad is not geolocalisate when he disable community mode, however on `/around/` he could enforce it with a button in a map controler (on a right of the map)

capture d ecran 2017-09-20 a 17 16 18

**Comment by newick** [comment]
the idea looks fine for me.
I'd add a link on the community part on the spy text.
The `i` icon is still no align with the text ;)
@newick I forgot, I change it now


An improvement could be: 
- as a writer I can share my content only if I send url but it's not available on search engines

**Comment by fuentesloic** [comment]
@newick I also increase the size, I think it better
capture d ecran 2017-09-20 a 17 45 45

**Comment by vinyll** [comment]
Man, you're coding faster than I can review!
I don't think that articles are a privacy issue anyhow. We may want to keep them public when they've been published. Not much purpose of writing a blog article otherwise.

**Comment by vinyll** [comment]
Man, you're coding faster than I can review!
I don't think that articles are a privacy issue anyhow. We may want to keep them public when they've been published. Not much purpose of writing a blog article otherwise.
I cannot review this these days. I'll do that next week :/

**Comment by fuentesloic** [comment]
Ok perfect :) 
What do you think about hidden writter datas in the header ?

File: templates/user/privacy.html [hunk 1]
```diff
@@ -1,41 +1,112 @@
...
+      </label>
+    </form>
+    <aside>
+      {% if user.allow_community %}
+      <p>
+        Disable community mode hidden your public data as publication, localisation and account.
+      </p>
+      {% else %}
+      <p>
+        Enable community mode allow you to share your articles, localisation and account.
```
Comment (vinyll) [inline templates/user/privacy.html]: Enabling community mode allows you to publish your articles, to share your location, and visitors to read your profile.

File: templates/user/privacy.html [hunk 2]
```diff
@@ -1,41 +1,112 @@
...
+    <form>
+      <input type=checkbox name=community {% if user.allow_community %}checked{% endif %} id=community>
+      <label for=community title="{% if user.allow_community %}Disable{% else %}Enable{% endif %} community profile">
+        Click to {% if user.allow_community %}disable{% else %}enable{% endif %}
+      </label>
+    </form>
+    <aside>
+      {% if user.allow_community %}
+      <p>
+        Disable community mode hidden your public data as publication, localisation and account.
```
Comment (vinyll) [inline templates/user/privacy.html]: Disabling the community mode hides your profile, your articles and your location.

File: templates/user/privacy.html [hunk 3]
```diff
@@ -1,41 +1,112 @@
...
+const communityField = document.querySelector('[name=community]')
+communityField.addEventListener('change', event => {
+  call = api.send('patch', "{{ url_for('profile_edit') }}",
+    {allow_community: event.target.checked},
+  )
+  call = api.send('patch', "{{ url_for('profile_edit') }}",
+    {allow_localization: event.target.checked},
+  )
+  if (event.target.checked) {
+    apiFeedback(call, 'You enable community mode, you share your informations, authorized contents and position.')
```
Comment (vinyll) [inline templates/user/privacy.html]: Community mode was enabled. Your profile, location and articles are now public.

File: templates/user/privacy.html [hunk 4]
```diff
@@ -1,41 +1,112 @@
...
+    localize()
+  } else {
+    apiFeedback(call, 'Your disable community mode, your hidden your profile and datas.')
```
Comment (vinyll) [inline templates/user/privacy.html]: Community mode was disabled. Your profile, location and articles are now hidden.

File: templates/user/profile.html [hunk 5]
```diff
@@ -19,34 +19,44 @@
...
+        onerror="this.src='{{ url_for('static', filename='images/avatar.png') }}'">
     </figure>
     <div>
       <h1 name=username>{{ user.username or 'Insert a username' }}</h1>
       <p name=about>{{ user.about|htmlnewline or 'Introduce yourself here' }}</p>
+      {% if user.allow_community %}
+      <a href={{ url_for_trips(user) }} data-latlng={{ user.current_location|join(',') }}></a>
+      {% else %}
+      <p class=private>
+        Community mode is disable.<br>
```
Comment (vinyll) [inline templates/user/profile.html]: Community mode is _disabled_; your contents are restricted and private.

File: templates/user/privacy.html [hunk 6]
```diff
@@ -1,41 +1,112 @@
...
+      {% else %}
+      <p>
+        Enable community mode allow you to share your articles, localisation and account.
+      </p>
+      {% endif %}
+    </aside>
+    {% if user.allow_community %}
+    <footer>
+      <p>
+        Even if you disable the community mode you can still read, write and use your maps.
```
Comment (vinyll) [inline templates/user/privacy.html]: Disabling the community mode allows only you to view your map.

File: templates/private.html [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+{% block extra_css %}
+<link rel=stylesheet href={{ url_for('static', filename='css/error.css') }}>
+{% endblock %}
+
+{% block main %}
+
+<section>
+  <h2>Don't worry,<br>
+    it's just a <span>{{ user.username }}</span> private page</h2>
```
Comment (vinyll) [inline templates/private.html]: this is just <span>{{ user.username }}</span>'s page that is private.</h2>

File: templates/private.html [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  <p>
+    Her/him don't want use community mode.<br>
```
Comment (vinyll) [inline templates/private.html]: <span>{{ user.username }}</span> has disabled the community mode.<br>

**Review by vinyll — COMMENTED** [review]
A few English sentences that can be improved.

`allow_location` has no more sense, we should remove it if `allow_community`.

I would have a different vision of the functionalities of the community mode.
The site in non community mode right now is barely  usable.

To me _community mode_ means that a user creates an account that is viewable by visitors, can publish articles that are readable by visitors (what's the purpose of publishing otherwise?). However, location are a privacy concern that a user might want to opt-in for. IMHO, that's what the _community mode_ is for, delibarately share one's location to share the trips you made and to be discoverable in your neightborhoods.

Also the _privacy mode_ should not permit Neomad to collect the user's position. Therefore privacy mode should disable the trips functionality.

**Comment by vinyll** [comment]
Suggestion of privacy mode vs community mode:

In community mode, all is viewable by others (except draft articles).

In private mode, the following changes apply:
- Neomad does not collect user's location
- Trips are not available (not even to the user as location is disabled)
- Around starts in a global view (we don't have the initial user location)
- Obviously the user won't be findable when searching a nomad around.

In private mode all the rest remains publicly readable: the user profile, the published articles. These data can be easily fine tuned by the user himself (adapting his bio, removing his photo, …)

File: templates/user/profile.html [hunk 9]
```diff
@@ -19,34 +19,44 @@
...
+        height=100
+        onerror="this.src='{{ url_for('static', filename='images/avatar.png') }}'">
     </figure>
     <div>
       <h1 name=username>{{ user.username or 'Insert a username' }}</h1>
       <p name=about>{{ user.about|htmlnewline or 'Introduce yourself here' }}</p>
+      {% if user.allow_community %}
+      <a href={{ url_for_trips(user) }} data-latlng={{ user.current_location|join(',') }}></a>
+      {% else %}
+      <p class=private>
```
Comment (vinyll) [inline templates/user/profile.html]: `<a class=private href={{ url_for('privacy') }}>…`  instead?

File: trips/views.py [hunk 10]
```diff
@@ -13,11 +13,15 @@
...
+        community = user.allow_community
     except User.DoesNotExist:
         abort(404)
-    return render_template('trips/map.html',
+    if community:
+        return render_template('trips/map.html',
                            user=user,
                            locations=user.locations)
+    else:
+        return render_template('private.html', user=user)
```
Comment (vinyll) [inline trips/views.py]: `return render_template('private.html', user=user), 403`

This is the status code for _Forbidden_

**Comment by fuentesloic** [comment]
@vinyll I agree with your comments!! about `allow_location`, I have a trouble if I remove it ... It's like the systeme is waiting after an `allow_location` in a profile (maybe in the DB?) so because it waiting after it that create a bug and a don't find out how handle it.

File: templates/user/profile.html (hunk 9 above)
Comment (fuentesloic) [reply templates/user/profile.html]: yeah it's a good idea !

File: trips/test.py [hunk 11]
```diff
@@ -2,11 +2,7 @@
 from unittest import TestCase
 
 from core import app
-from blog import views
```
Comment (vinyll) [inline trips/test.py]: All these are required in this file

File: blog/test.py [hunk 12]
```diff
@@ -148,8 +143,8 @@ def test_content_cleaned_html(self):
                           content='<p><img src="/static/img"></p>').save()
         self.assertEqual(article.content, '<p><img src="/static/img"/></p>')
         article = Article(title='',
-                          content='<p><em>emphased content</em><i>also emphased</i></p>').save()
-        self.assertEqual(article.content, '<p><em>emphased content</em><em>also emphased</em></p>')
+                          content='<p><i>also emphased</i></p>').save()
```
Comment (vinyll) [inline blog/test.py]: We should never have `<i>` tags

**Review by vinyll — CHANGES_REQUESTED** [review]

**Comment by fuentesloic** [comment]
@vinyll I think it's the last step ... BUT
I dunno understand why my test doesn't work ... it's the same that you write in master, I just change variation according allow_community.
The bug is from test about the adress Email, could you check quickly, because I try but it's still not working.
___Fuck****ing test___ :)

**Comment by fuentesloic** [comment]
@vinyll Still need help for test :(

**Comment by fuentesloic** [comment]
@vinyll do we push that on dev to test ??

**Review by vinyll — APPROVED** [review]

**Comment by vinyll** [comment]
Shall you merge @fuentesloic so we preprod-deploy? ;)

**Pull request merged**